File: src/main/java/com/github/samarium150/structurescompass/gui/StructureSearchList.java
package com.github.samarium150.structurescompass.gui;

import com.github.samarium150.structurescompass.util.RenderUtils;
import com.github.samarium150.structurescompass.util.StructureUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.widget.list.ExtendedList;
import net.minecraft.world.gen.feature.structure.Structure;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import java.util.HashMap;
import java.util.Objects;

/**
 * Class of structures for searching
 * <p>
 * This class is adapted from
 * <a href="https://github.com/MattCzyr/NaturesCompass" target="_blank">NaturesCompass</a>
 * which is under
 * <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" target="_blank">
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
 * </a>.
 */
@OnlyIn(Dist.CLIENT)
public final class StructureSearchList extends ExtendedList<StructureSearchEntry> {
    
    private final StructuresCompassScreen screen;
    private final HashMap<String, StructureSearchEntry> map = new HashMap<>();
    
    public StructureSearchList(StructuresCompassScreen screen, Minecraft minecraft,
                               int widthIn, int heightIn, int topIn, int bottomIn, int slotHeightIn) {
        super(minecraft, widthIn, heightIn, topIn, bottomIn, slotHeightIn);
        this.screen = screen;
        refresh();
    }
    
    public StructuresCompassScreen getScreen() {
        return screen;
    }
    
    public void selectStructure(StructureSearchEntry entry) {
        setSelected(entry);
        screen.selectStructure(entry);
    }
    
    public void selectStructure(Structure<?> structure) {
        selectStructure(map.get(StructureUtils.getStructureName(structure)));
    }
    
    public void refresh() {
        clearEntries();
        for (Structure<?> structure : screen.sortStructures()) {
            StructureSearchEntry entry = new StructureSearchEntry(this, structure);
            addEntry(entry);
            map.put(StructureUtils.getStructureName(structure), entry);
        }
        selectStructure((StructureSearchEntry) null);
        setScrollAmount(0D);
        changeFocus(false);
    }
    
    public void restoreFocus() {
        if (getSelected() != null)
            changeFocus(true);
    }
    
    @Override
    protected int getScrollbarPosition() {
        return super.getScrollbarPosition() + 20;
    }
    
    @Override
    public int getRowWidth() {
        return super.getRowWidth() + 50;
    }
    
    @Override
    protected boolean isSelectedItem(int index) {
        return index >= 0 && index < children().size() && children().get(index).equals(getSelected());
    }
    
    @Override
    protected void renderList(int x, int y, int mouseX, int mouseY, float partialTicks) {
        for (int i = 0; i < getItemCount(); ++i) {
            int top = getRowTop(i);
            int bottom = top + itemHeight;
            if (bottom + height >= y0 && top <= y1) {
                StructureSearchEntry entry = getEntry(i);
                int h = itemHeight - 4;
                int w = getRowWidth();
                if (isSelectedItem(i)) {
                    final int insideLeft = x0 + width / 2 - getRowWidth() / 2 + 2;
                    RenderUtils.drawRect(insideLeft - 4, top - 4, insideLeft + getRowWidth() + 4, top + itemHeight, 255 / 2 << 24);
                }
                int left = getRowLeft();
                entry.render(
                    i, top, left, w, h, mouseX, mouseY,
                    isMouseOver(mouseX, mouseY) && Objects.equals(getEntryAtPosition(mouseX, mouseY), entry),
                    partialTicks
                );
            }
        }
    }
    
    @Override
    public void render(int mouseX, int mouseY, float partialTicks) {
        renderList(getRowLeft(), y0 + 4 - (int) getScrollAmount(), mouseX, mouseY, partialTicks);
    }
}

File: src/main/kotlin/io/github/samarium150/minecraft/mod/structures_compass/client/gui/widget/StructureSearchList.kt
/*
 * Copyright (c) 2022 Samarium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
 */
package io.github.samarium150.minecraft.mod.structures_compass.client.gui.widget

import io.github.samarium150.minecraft.mod.structures_compass.client.gui.StructuresCompassScreen
import io.github.samarium150.minecraft.mod.structures_compass.client.util.drawRect
import io.github.samarium150.minecraft.mod.structures_compass.util.getIdentifier
import io.github.samarium150.minecraft.mod.structures_compass.util.minecraftClient
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.gui.widget.AlwaysSelectedEntryListWidget
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.Identifier
import net.minecraft.world.gen.feature.StructureFeature
import java.util.*

@Environment(EnvType.CLIENT)
class StructureSearchList(
    val screen: StructuresCompassScreen,
    width: Int,
    height: Int,
    top: Int,
    bottom: Int,
    slotHeight: Int
) : AlwaysSelectedEntryListWidget<StructureSearchEntry>(
    minecraftClient, width, height, top, bottom, slotHeight
) {
    private val map = mutableMapOf<Identifier, StructureSearchEntry>()

    init {
        refresh()
    }

    override fun getScrollbarPositionX(): Int {
        return super.getScrollbarPositionX() + 20
    }

    override fun getRowWidth(): Int {
        return super.getRowWidth() + 50
    }

    override fun isSelectedEntry(slotIndex: Int): Boolean {
        return if (slotIndex >= 0 && slotIndex < children().size) children()[slotIndex].equals(selectedOrNull) else false
    }

    private fun getRowBottom(entryIndex: Int): Int {
        return getRowTop(entryIndex) + itemHeight
    }

    override fun renderList(
        matrixStack: MatrixStack?,
        x: Int,
        y: Int,
        mouseX: Int,
        mouseY: Int,
        partialTicks: Float
    ) {
        for (i in 0 until entryCount) {
            val rowTop = getRowTop(i)
            val rowBottom = getRowBottom(i)
            if (rowBottom + height >= top && rowTop <= bottom) {
                val entry = getEntry(i)
                val h = itemHeight - 4
                val w = rowWidth
                if (isSelectedEntry(i)) {
                    val insideLeft: Int = left + width / 2 - rowWidth / 2 + 2
                    drawRect(
                        insideLeft - 4,
                        rowTop - 4,
                        insideLeft + rowWidth + 4,
                        rowTop + itemHeight,
                        255 / 2 shl 24
                    )
                }
                entry.render(
                    matrixStack, i, rowTop, rowLeft, w, h, mouseX, mouseY,
                    isMouseOver(mouseX.toDouble(), mouseY.toDouble()) &&
                        Objects.equals(getEntryAtPosition(mouseX.toDouble(), mouseY.toDouble()), entry),
                    partialTicks
                )
            }
        }
    }

    override fun render(matrixStack: MatrixStack?, mouseX: Int, mouseY: Int, partialTicks: Float) {
        renderList(matrixStack, rowLeft, top + 4 - scrollAmount.toInt(), mouseX, mouseY, partialTicks)
    }

    fun selectStructure(entry: StructureSearchEntry?) {
        setSelected(entry)
        screen.selectStructure(entry)
    }

    fun selectStructure(structure: StructureFeature<*>) {
        selectStructure(map[structure.getIdentifier()])
    }

    fun refresh() {
        clearEntries()
        for (structure in screen.sortStructures()) {
            val entry = StructureSearchEntry(this, structure)
            addEntry(entry)
            map[structure.getIdentifier()!!] = entry
        }
        selectStructure(null)
        scrollAmount = 0.0
        changeFocus(false)
    }

    fun restoreScrollAmount() {
        if (selectedOrNull != null) {
            changeFocus(true)
            scrollAmount += itemHeight * 2
        }
    }
}
